onnxruntime/contrib_ops/cpu/bert/qembed_layer_norm.cc

```diff
@@ -0,0 +1,243 @@
+            failed.store(true, std::memory_order_release);
+            return;
+          }
+          int position_col_index = index % sequence_length;
+          if (position_col_index >= position_embedding_length) {
+            failed.store(true, std::memory_order_release);
+            return;
+          }
+          int segment_col_index = 0;
+          if (nullptr != segment_ids_data) {
+            segment_col_index = segment_ids_data[index];
+            if (segment_col_index < 0 || segment_col_index >= segment_embedding_length) {
+              failed.store(true, std::memory_order_release);
+              return;
+            }
+          }
+
+          // Grab inputs for the embeddings for the current batch index:
+          const uint8_t* input_word_embedding = word_embedding_data + word_col_index * hidden_size;
+          const uint8_t* input_position_embedding =
+              position_embedding_data + position_col_index * hidden_size;
+          const uint8_t* input_segment_embedding = nullptr;
+          if (segment_embedding_data != nullptr) {
+            input_segment_embedding = segment_embedding_data + segment_col_index * hidden_size;
+          }
+
+          T* output = output_data + (index * hidden_size);
+
+          T sum = static_cast<T>(0);
+          for (int i = 0; i < hidden_size; ++i) {
+            T subtotal = Dequantize<uint8_t>(input_word_embedding[i],
+                                             word_embedding_scale,
+                                             word_embedding_zero_point) +
+                         Dequantize<uint8_t>(input_position_embedding[i],
+                                             position_embedding_scale,
+                                             position_embedding_zero_point);
+            if (segment_embedding_data != nullptr) {
+              subtotal += Dequantize<uint8_t>(input_segment_embedding[i],
+                                              segment_embedding_scale,
+                                              segment_embedding_zero_point);
+            }
+            output[i] = subtotal;
+            sum += subtotal;
+          }
+
+          T mean = sum / hidden_size;
+          sum = 0;
+
+          for (int i = 0; i < hidden_size; i++) {
+            T a = output[i] - mean;
+            output[i] = a;
+            sum += a * a;
+          }
+
+          T e = sqrt(sum / hidden_size + epsilon_);
+          for (int i = 0; i < hidden_size; i++) {
+            T cur_gamma = Dequantize<uint8_t>(gamma_data[i],
+                                              layer_norm_weights_scale,
+                                              layer_norm_weights_zero_point);
+            T cur_beta = Dequantize<uint8_t>(beta_data[i],
+                                             layer_norm_bias_scale,
+                                             layer_norm_bias_zero_point);
+            output[i] = output[i] / e * cur_gamma + cur_beta;
```

Comment:
Is quantization of gamma_data and beta_data paid off? They are quite small.

Reply:
The models we see have a large layer (768). I'd like to stick with them for now and use Prepack() to help with this in the future.